app/js/arethusa.search/services/search.js

```diff
@@ -40,6 +40,182 @@ angular.module('arethusa.search').service('search', [
       });
     };
 
+    this.findWordInContext = function(query) {
+
+      return arethusaUtil.inject([], self.strings, function (memo, string, ids) {
+        var matches = []
+        angular.forEach(ids, function(id) {
+          var previousToken = state.getPreviousTokens(id,1);
+          var nextToken = state.getNextTokens(id,1);
+          var nextString = nextToken.length > 0 ? nextToken[0].string : null;
+          var prevString = previousToken.length > 0 ? previousToken[0].string : null;
+          var match = self.compareWordsWithContext(string,prevString,nextString,query.word);
+          if (match.match) {
+            var matchData = {
+              id: id,
+              matchedPrefix: 0,
+              matchedSuffix: 0
+            }
+            if (match.combine < 0) {
+              matchData.includePrevious = previousToken[0].id;
+            } else if (match.combine > 0) {
+              matchData.includeNext = nextToken[0].id;
+            }
+            matches.push(matchData)
+          }
+        });
+        // we only need to try to further narrow by context if we got more than
+        // one hit
+        if (matches.length > 1) {
+          matches.forEach(function (match) {
+            var startingId = match.includePrefix ? match.includePrefix :
+              match.includeSuffix ? match.includeSuffix : match.id;
+            var prefixWords = tokenize(query.prefix);
+            var suffixWords = tokenize(query.suffix);
+            var matchedPrefixWords = 0;
+            var matchedSuffixWords = 0;
+            var previousTokens = [];
+            if (prefixWords.length > 0) {
+              var previousTokens = state.getPreviousTokens(startingId);
+              var tokenIndex = previousTokens.length-1;
+              for (var i=prefixWords.length-1; i>=0; i--) {
+                if (previousTokens[tokenIndex]) {
+                  var token = previousTokens[tokenIndex];
+                  var nextString = previousTokens[tokenIndex+1] ? previousTokens[tokenIndex+1].string : null;
+                  var prevString = tokenIndex > 0 && previousTokens[tokenIndex-1] ? previousTokens[tokenIndex-1].string : null;
+                  var matchP = self.compareWordsWithContext(token.string,prevString,nextString,prefixWords[i]);
+                  if (matchP.match) {
```

Comment:
If it finds no match should the algorithm give up or continue matching the rest of the `prefixWords`?

Reply:
I want it to keep trying to match. I'm thinking about the case where maybe the input string is missing a word or two from the treebank or vice versa.  I started with trying to do an exact match on prefix and suffix requested, but it got very convoluted and I'm not sure it's worth it given that there are going to be variations in tokenization as well as enclytics, ellipses, punctuation etc.  If we were searching more than a single sentence then this would have to be more exact, but within a single sentence, I'm thinking that maybe it will be sufficient to just find the best match -- i.e the one with the greatest number of matching words from the prefix and suffix string.  The likelihood of multiple occurrences of the same word having an equivalent number of matches in the prefix and suffix string seems like it might be small enough to suffice. This needs more testing with real data of course.

Reply:
That makes sense, and the algorithm can be iterated on after it's merged.